### ISSUE TITLE
CVC4 Support

### ISSUE THREAD

**Issue by cpitclaudel** [FStarLang/FStar#1086]
It would be nice to actually support CVC4. `--help` says `--smt [path]  Path to the SMT solver (usually Z3, but could be any SMT2-compatible solver)`, but the code strongly assumes Z3 (to uses CVC4, one would among other things have to start it with `--lang smt`, for example).

(I recently chatted with Cesare at MIT, and CVC4 seems to have lots of neat features)

**Comment by catalin-hritcu** [comment]
As you noticed, our current reliance of Z3 is very strong, and I think that's very unlikely to change any time soon, since predictability is **much** more important to us than solving more SMT queries. What would be more realistic, would be to allow **failing** proof obligations to also be tried with CVC4 or whatever, just in case it manages to solve them. If this happens we would still fail, but at least the user would get a better idea that the obligation is actually true. So the use case I could envision is similar to the use of Z3 parallelism for getting more information on which queries are actually true, while still failing if the Z3 answer is not predictable: see https://github.com/FStarLang/FStar/issues/889#issuecomment-286729857

**Comment by catalin-hritcu** [comment]
Just for the record, last time I talked with Nik about this it was in the context of a discussion on Why3, which is another great tool that can multiplex between multiple FOL solvers, not just SMT. So a more general way to integrate CVC4 and other solvers would be by hooking into the Why3 solver backend, like EasyCrypt does. At least for the use case I outline above, it could be interesting.

For now, I've simply made it clearer that currently Z3 is the only supported SMT solver (ba94f8330).
```
  --smt [path]  Path to the Z3 SMT solver (currently Z3 is the only supported solver)
```

**Comment by cpitclaudel** [comment]
> As you noticed, our current reliance of Z3 is strong, and I think that's unlikely to change any time soon, since predictability is much more important to us than solving more SMT queries. 

Do we have evidence than Z3 is more stable than CVC4? (I don't have enough experience with either to say)

**Comment by catalin-hritcu** [comment]
> Do we have evidence than Z3 is more stable than CVC4?
> (I don't have enough experience with either to say)

That's not what I was trying to say. What I meant is that verifying our code (libraries, examples, everything) depends on Z3 and can't expect to change to a different SMT solver and still verify the code. We don't even support different Z3 versions ...

**Comment by cpitclaudel** [comment]
> That's not what I was trying to say. What I meant is that verifying our code (libraries, examples, everything) depends on Z3 and can't expect to change to a different SMT solver and still verify the code. We don't even support different Z3 versions ...

Makes sense. Assuming the stdlib verifies (or is made to verify) with CVC4 as well, though, I could imagine other F* projects using it exclusively instead of Z3.  And anything where CVC4 is significantly less predictable than Z3 would probably be viewed as a bug by the CVC4 folks.

**Comment by cpitclaudel** [comment]
I just wrote to the CVC4 mailing list to ask for tips.  Here's a summary:

* F* doesn't issue a `set-logic` call (we should use `(set-logic ALL)`, IIUC)

* F* uses options that CVC4 doesn't know about:

    - `(set-option :global-decls false)`
      This should be the default behaviour, so nothing to do here.  Why do we even have this?

    - `(set-option :smt.mbqi false)`
      CVC4 has a command-line flag for this

    - (set-option :auto_config false)
      Probably doesn't matter

* F* calls `declare-sort` with a single argument (`(declare-sort FString)`); it should be changed to `(declare-sort FString 0)`.

* F* uses `implies` and `iff`.  These seem to be non-standard; can we change them to `=>` and `=`?

* F* uses :qid annotations on assertions.  What are these for?

* F* sometimes adds `:weight 0` annotations on quantifiers.  Not sure what the CVC4 equivalent is.

* F* starts some names with `@` (e.g. `(assert (… :named @fuel_correspondence_Prims.pow2.fuel_instrumented))`), but names starting with `@` are reserved in SMTLIB.  Can we change the naming scheme to e.g. `__fuel_correspondence_Prims.pow2.fuel_instrumented`?

* F* uses `(eval` to fetch values from models.  CVC4 doesn't have that; I asked about it.

Applying all the fixes above and running CVC4 on the modified input works, but I'm running into performance issues.  I have an example that verifies in roughly 12 seconds with Z3 on my machine, whereas it takes over 15 minutes in CVC4 (I killed it after letting `cvc4 --incremental --e-matching --mbqi=none --lang smt queries-STLC.smt2` run for that long).

**Comment by prosecco** [comment]
I’ve tried CVC4 on some HACL specs. For bitvector and math, it seems to work *much* better than Z3.
Indeed, as Catalin points out below, the scripts we generate from F* are not compatible with CVC4.
I would love to generate generic SMT2. JK and I have been playing with this a bit and would be happy
to help design and/or test a more generic SMT backend.

-Karthik

**Comment by cpitclaudel** [comment]
> I’ve tried CVC4 on some HACL specs. For bitvector and math, it seems to work *much* better than Z3.

Exciting!

> I would love to generate generic SMT2. JK and I have been playing with this a bit and would be happy to help design and/or test a more generic SMT backend.

I think the steps are pretty simple:

* **Fix the command line invocation**:  Right now we hardcode `-smt2 -in auto_config=false model=true smt.relevancy=2`.  Instead, we should pass different options depending on the smt solver. For CVC4, the proper option is probably something like `--incremental --mbqi=none --lang smt2`

* **Fix calls to declare-sort**: It's missing a 0

*  **Use `=>` and `=` instead of `implies` and `iff`**

* **Use `__` instead of `@` as a prefix for quantifier names.

* **Find a replacement for eval**: Hopefully someone on the CVC4 mailing list can help.  Using get-model would be one way to go, but then we'd have to make sure to not call it if we get unsat, since in CVC4 that's a fatal error.

@prosecco, @jkzinzindohoue, do you want to try these?  The CVC-users thread is at http://cs.nyu.edu/pipermail/cvc-users/2017/000890.html

Clément.

**Comment by cpitclaudel** [comment]
Quick update: Andrew Reynolds tell me that we should probably be using get-values instead of eval, and that they'll consider changing the error that it returns when no model is available to non-fatal.  With this solved, the only blockers will be on the F* side :)

**Comment by nbraud** [comment]
@cpitclaudel Following my conference call with @nikswamy and @wintersteiger, I agreed to put in some work towards having a SMTLIB-compliant SMT encoding, which should let you use CVC4  :)

**Comment by cpitclaudel** [comment]
Very nice! Let me know if you need info/help.

Once we have a working encoding, there'd be plenty of fun avenues to explore, too, if anyone's curious :) 

- CVC4 has a theory of finite relations (in CAV 2017) and of finite sets (2016) which could be used to encode parts of the type system with less quantifiers
- CVC4 has a theory of strings (I think Z3 does too, though I'm nto sure it includes utf-8), and a theory of regular expressions.
- The latest SMTLib standard has a construct called `define-fun-rec` which lets you define recursive functions without explicit quantifiers, and CVC4 has special support for them (though mostly in the context of finite model finding, IIUC).

**Comment by nbraud** [comment]
@cpitclaudel Interestingly, I've also been wanting to use more Z3's builtin theories.
However, @nikswamy pointed out -- correctly -- that this gets quite tricky when we have polymorphic functions to deal with; I will check what Leon does in those cases, for inspiration.

**Comment by nbraud** [comment]
@cpitclaudel I took the liberty to edit your earlier comment (fixing the formatting, adding checkboxes to track progress, ...).

Something which might be a pain are solver options; the easy way out would be to push them into the Z3 CLI arguments (i.e. change the default value of `--z3cliopt`), and forcing you (or rather, anyone choosing to use a different SMT solver) to set different options (`--z3cliopt ''` in the extreme case).

**Comment by cpitclaudel** [comment]
Thanks!

> Something which might be a pain are solver options; the easy way out would be to push them into the Z3 CLI arguments (i.e. change the default value of --z3cliopt), and forcing you (or rather, anyone choosing to use a different SMT solver) to set different options (--z3cliopt '' in the extreme case).

We could just detect the name of the binary: if it ends in `z3` or `z3.exe`, then it gets the z3 options; and if it ends in `cvc4` or `cvc4.exe` it gets the CVC4 options.  

In any case, I think CVC4 ignores unknown options, so maybe that's our easy way out?

**Comment by nbraud** [comment]
@cpitclaudel I actually introduced a `solver` datatype, and produce different CLI options based on that.
This is in branch `nbraud/smtlib`, with a bunch of related changes, but it's nowhere complete for CVC4 support (PR pending soon).

**Comment by cpitclaudel** [comment]
Sounds great! Do you think this would be a good occasion to do the but of refactoring that was discussed earlier? I'm trying to get F* running in the browser, and it's much easier for me to plug in a JavaScript (transpiled) version of CVC4 if there's a minimal interface that I need to implement (say with 3 methods, init, eval_query, and teardown). Would that be hard?

**Comment by wintersteiger** [comment]
That reminds me that I haven't answered your question @cpitclaudel !

It's possible but may require us to move/rewrite/duplicate code to produce a stripped down version of the current Z3 interface (which supports lots of stuff that won't be needed in this case). I think the best way forward for this is to implement `ask` functions like the one in FStar.SMTEncoding.Z3.fs in new files/modules. Those functions can then use the theory filter etc, from FStar.SMTEncoding.Solver.fs, but they don't need to implement `run_job` or `enqueue`. 

The callback are a bit tricky, but I think those would continue to live in the F* world. Let me know if you need more information about those.

**Comment by cpitclaudel** [comment]
> It's possible but may require us to move/rewrite/duplicate code to produce a stripped down version of the current Z3 interface (which supports lots of stuff that won't be needed in this case).

Makes sense!  I think I could probably make simple shims for some of them, too, thus reducing the duplication needed.  Maybe a good first step would be to isolate the Z3 process manipulation stuff to a separate file to get a clear picture of what the current (implicit) interface? They I can see how much of it I can provide or shim out

> The callback are a bit tricky, but I think those would continue to live in the F* world. Let me know if you need more information about those.

This would be very useful.  Does F* ever explicitly wait on a process/callback?  IIRC the interface is synchronous, so I guess it thread-joins on the various solvers instances at some point?  Presumably I can make this work by running synchronously in the js case, and then the call to `join` would return immediately.

**Comment by wintersteiger** [comment]
Yes, that's one of the complications in the current implementation; it supports synchronous and asynchronous queries and in the synchronous case it supports incremental solver calls as well. All the state required for those things is kept in the call-backs, which can be a bit confusing. 

So, for the JS case, support only synchronous calls and only non-incremental ones (assume --z3refresh is set). For the asynchronous case there's the `finish` function which will join the threads, but for JS we really shouldn't need any of that. The synchronous parts wait right there and then for the results (in `doZ3Exe'`).

**Comment by cpitclaudel** [comment]
> Yes, that's one of the complications in the current implementation; it supports synchronous and asynchronous queries and in the synchronous case it supports incremental solver calls as well. All the state required for those things is kept in the call-backs, which can be a bit confusing.

This makes sense; thanks for the explanation!  Callbacks are not too too much of an issue, and we could even have multiple solver instances running (using multiple web workers).  The one significant issue with that model is that JS doesn't have a notion of joining: it's all synchronous, or all asynchronous, so you can't start a bunch of workers then join on them (see e.g. https://stackoverflow.com/questions/9242512/is-it-possible-to-cause-web-workers-to-sleep).  So I agree that synchronous goals are the way to go atm.

> So, for the JS case, support only synchronous calls and only non-incremental ones (assume --z3refresh is set). For the asynchronous case there's the finish function which will join the threads, but for JS we really shouldn't need any of that. The synchronous parts wait right there and then for the results (in doZ3Exe').

Neat, sounds like a plan. I think it wouldn't be much harder to add incremental support after that, since that can still be synchronous (right?). It would require small patches to the solvers, but nothing huge, I think.

Thanks!

**Comment by cpitclaudel** [comment]
> The synchronous parts wait right there and then for the results (in doZ3Exe').

I'll look at this part of the file in more details, then, to get started on this.

**Comment by cpitclaudel** [comment]
@nbraud CVC4 was just updated to not exit when get-value or get-model is called in an unsat context: https://github.com/CVC4/CVC4/commit/54d24c786d6a843cc72dfb5e377603349ea5e420

**Comment by nbraud** [comment]
@cpitclaudel Thanks a bunch.
I am still hammering on a couple of issues on my side, but #1099 should soon be mergeable.

**Comment by cpitclaudel** [comment]
Nice, thanks!

**Comment by nikswamy** [comment]
@nbraud @wintersteiger : any updates on this?

**Comment by catalin-hritcu** [comment]
Current status is that the corresponding PR was closed for now: "I'm closing this PR. It would be great to support non-Z3 solvers. So, if someone wants to revive it, that would be great."
https://github.com/FStarLang/FStar/pull/1099#issuecomment-518078929
